src/solc_0.8/estate/EstateBaseToken.sol

```diff
@@ -111,30 +118,78 @@ contract EstateBaseToken is ImmutableERC721, Initializable {
         _addQuads(sender, estateId, sizes, xs, ys, junctions, false);
     }
 
-    function destroy(address sender, uint256 estateId) external {
+    /// @notice Burns token `id`.
+    /// @param id The token which will be burnt.
+    function burn(uint256 id) external override {
+        address sender = _msgSender();
         _check_authorized(sender, BREAK);
-        _check_hasOwnerRights(sender, estateId);
-        _owners[estateId] = 0; // TODO keep track of it so it can transfer Land back
-        _numNFTPerAddress[sender]--;
-        emit Transfer(sender, address(uint160(0)), estateId);
+        _check_hasOwnerRights(sender, id);
+        _burn(sender, _ownerOf(id), id);
+    }
+
+    /// @notice Burn token`id` from `from`.
+    /// @param from address whose token is to be burnt.
+    /// @param id The token which will be burnt.
+    function burnFrom(address from, uint256 id) external override {
+        require(from != address(uint160(0)), "NOT_FROM_ZERO_ADDRESS");
+        (address owner, bool operatorEnabled) = _ownerAndOperatorEnabledOf(id);
+        require(owner != address(uint160(0)), "NONEXISTENT_TOKEN");
+        address msgSender = _msgSender();
+        require(
+            msgSender == from ||
+                (operatorEnabled && _operators[id] == msgSender) ||
+                _superOperators[msgSender] ||
+                _operatorsForAll[from][msgSender],
+            "UNAUTHORIZED_BURN"
+        );
+        _burn(from, owner, id);
+    }
+
+    /// @notice Update an existing ESTATE token.This actually burns old token
+    /// and mints new token with same basId & incremented version.
+    /// @param from The one updating the ESTATE token.
+    /// @param estateId The current id of the ESTATE token.
+    /// @param ids The ids to use for the update.
+    /// @param junctions The junctions to use for the update.
+    /// @return The new estateId.
+    function updateEstate(
+        address from,
+        uint256 estateId,
+        uint256[] memory ids,
+        uint256[] memory junctions,
+        bytes32 uri
+    ) external returns (uint256) {
+        // @review can this function also handle removing lands?
```

Comment:
it might be possible
but there might be too many lands at this point that even if we gave the full list to the contract, it would run out of gas to check their adjacency.

Reply:
Yes, As per the google doc spec conversation, this would have to involve a complete rebuild of the token (checking adjacency, etc).